File: Python/Problems/problem_1.py
#!/usr/bin/env python3

from Utils.misc import sum_range

def solution():
    UPPER = 1000
    mult_sum = sum_of_multiples_of_3_and_5(UPPER)
    return f'The sum of multiples of 3 and 5 below {UPPER} is {mult_sum}'


def sum_multiples_of_3_and_5(UPPERBOUND):
    UPPERBOUND -= 1
    sum_mult = 5 * sum_range(UPPERBOUND // 5)
    sum_mult += 3 * sum_range(UPPERBOUND // 3)
    sum_mult -= (5 * 3) * sum_range(UPPERBOUND // (5 * 3))

    return sum_mult


<<<<<<< HEAD:Problems/Problem_1/Python/problem_1.py
def sum_range(upper):
    sum_vals = ((upper + 1) // 2) * upper

    if not upper % 2:
        sum_vals += upper / 2

    return sum_vals


if __name__ == '__main__':
    sum_multiples_of_3_and_5(1000)

=======
>>>>>>> refactorProblemRunners:Python/Problems/problem_1.py

File: Rust/src/problems/problem_1.rs
use utils::misc::sum_range;


pub fn solution() -> String {
    let upper = 1000;
    let sum = sum_multiples_of_3_and_5(upper);
    format!("The sum of multiples of 3 and 5 below {} is {}", upper, sum)
}


fn sum_multiples_of_3_and_5(upper: usize) -> usize {
    let adj_upper = upper - 1;
    let mut sum = 5 * sum_range(adj_upper / 5);
    sum += 3 * sum_range(adj_upper / 3);
    sum -= 15 * sum_range(adj_upper / 15);

    sum
}


#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_sum_multiples_of_3_and_5() {
        assert_eq!(sum_multiples_of_3_and_5(10), 23);
    }
}

